File: src/ball.js
import { velocityUnit } from "./attributes.js";

// set_ball_speed (0x108e0): throw controls the time before velocity steps down.
export function setBallSpeed(ball, attribute) {
  ball.speedTimer = attribute >>> 1;
  ball.nextSlowdown = (attribute >>> 2) + (attribute >>> 3) + (attribute >>> 4);
  ball.slowFraction = 0;
}
export function slowBallFrame(ball) {
  if (ball.owner >= 0 || (!ball.vx && !ball.vz)) return;
  // Original longitudinal coordinates 48..1104, centered at 576.
  const terrainY = Math.round(576 - ball.x / (22.4 / 576));
  if (ball.vx && (terrainY < 48 || terrainY > 1104)) return;
  const timer = ball.speedTimer ?? 0;
  const next = ball.nextSlowdown ?? 0;
  if (timer === 0 || next > timer) {
    if (timer > 0) ball.nextSlowdown = (next >>> 1) + (next >>> 2);
    for (const key of ["vx", "vz"]) {
      const value = ball[key];
      let reduced =
        Math.sign(value) * Math.max(0, Math.abs(value) - velocityUnit);
      if (Math.abs(reduced) < 1e-9) reduced = 0;
      // step_slow_ball preserves the final velocity unit while airborne.
      if (reduced !== 0 || ball.h <= 0.25) ball[key] = reduced;
    }
  }
  if (timer > 0) ball.speedTimer = timer - 1;
}
export function slowBall(ball, dt) {
  ball.slowFraction = (ball.slowFraction ?? 0) + dt * 25;
  while (ball.slowFraction >= 1 - 1e-9) {
    ball.slowFraction = Math.max(0, ball.slowFraction - 1);
    slowBallFrame(ball);
  }
}

export function reflectBall(ball, axis) {
  const velocity = axis === "x" ? "vx" : "vz";
  const direction = axis === "x" ? "dirX" : "dirZ";
  ball[velocity] = -ball[velocity];
  if (ball[direction] !== undefined) ball[direction] = -ball[direction];
  if (axis === "x") ball.speedTimer = Math.ceil((ball.speedTimer ?? 0) / 2);
}
export function warpBall(ball, attribute) {
  let dx = ball.dirX,
    dz = ball.dirZ;
  if (dx === undefined || dz === undefined) {
    const angle =
      (Math.round(Math.atan2(ball.vz, ball.vx) / (Math.PI / 4)) * Math.PI) / 4;
    dx = Math.round(Math.cos(angle));
    dz = Math.round(Math.sin(angle));
  }
  ball.z = -Math.sign(ball.z) * 11.2;
  const drift = ball.vz;
  ball.vx = dx * 8 * velocityUnit;
  ball.vz = dz === 0 ? drift : dz * 8 * velocityUnit;
  ball.dirX = dx;
  ball.dirZ = dz;
  setBallSpeed(ball, attribute);
}

// throwing_action_fn 0x10882: only a zero component can receive release input.
export function steerRelease(ball, input = {}) {
  if (ball.vz === 0) ball.vz = Math.sign(input.z || 0) * 4 * velocityUnit;
  else if (ball.vx === 0) ball.vx = Math.sign(input.x || 0) * 4 * velocityUnit;
}

// Amiga memory 0x6c32 / 0x6c94; -2 terminator holds the final ground stage.
export const highFlight = [
  1, 1, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6,
  6, 5, 5, 5, 5, 5, 5, 4, 4, 4, 4, 4, 3, 3, 3, 3, 2, 2, 2, 1, 1, 0,
];
// Amiga anim_ball_launch at 0x6d8a; final -2 holds the last stage.
export const launchFlight = [0,7,7,8,8,9,9,10,10,6,6,6,6,6,6,6,6,6,6,6,6,6,6,10,10,9,9,8,8,7,7,0];
export const lowFlight = [...Array(16).fill(1), 0];
export function startFlight(b, high) {
  b.flightKind = high ? 2 : 1;
  b.flightIndex = -1;
  b.flightFraction = 0;
  b.flightStage = 1;
  b.h = 0.75;
  b.vh = 0;
}
export function flightStep(b, dt) {
  if (!b.flightKind) return false;
  b.flightFraction += dt * 25;
  const frames = b.flightKind === 3 ? launchFlight : b.flightKind === 2 ? highFlight : lowFlight;
  while (b.flightFraction >= 1 - 1e-9) {
    b.flightFraction = Math.max(0, b.flightFraction - 1);
    b.flightIndex = Math.min(b.flightIndex + 1, frames.length - 1);
    b.flightStage = frames[b.flightIndex];
  }
  // Launch sprites 7..10 depict the near-deck ascent/descent.
  const heightStage = b.flightKind === 3 && b.flightStage >= 7 ? b.flightStage - 6 : b.flightStage;
  b.h = 0.25 + heightStage * 0.5;
  b.vh = 0;
  return true;
}

File: server/ball.go
package main

import "math"

// set_ball_speed (0x108e0): throw controls the time before velocity steps down.
func setBallSpeed(b *Ball, attribute int) {
	b.SpeedTimer = attribute >> 1
	b.NextSlowdown = (attribute >> 2) + (attribute >> 3) + (attribute >> 4)
	b.SlowFraction = 0
}
func slowBallFrame(b *Ball) {
	if b.Owner >= 0 || (b.VX == 0 && b.VZ == 0) {
		return
	}
	if b.VX != 0 && math.Abs(b.X) > 528*22.4/576 {
		return
	}
	if b.SpeedTimer == 0 || b.NextSlowdown > b.SpeedTimer {
		if b.SpeedTimer > 0 {
			b.NextSlowdown = (b.NextSlowdown >> 1) + (b.NextSlowdown >> 2)
		}
		for _, v := range []*float64{&b.VX, &b.VZ} {
			reduced := math.Copysign(math.Max(0, math.Abs(*v)-velocityUnit), *v)
			if math.Abs(reduced) < 1e-9 {
				reduced = 0
			}
			if reduced != 0 || b.H <= .25 {
				*v = reduced
			}
		}
	}
	if b.SpeedTimer > 0 {
		b.SpeedTimer--
	}
}
func slowBall(b *Ball, dt float64) {
	b.SlowFraction += dt * 25
	for b.SlowFraction >= 1-1e-9 {
		b.SlowFraction = math.Max(0, b.SlowFraction-1)
		slowBallFrame(b)
	}
}

func reflectBall(b *Ball, longitudinal bool) {
	if longitudinal {
		b.VX = -b.VX
		b.DirX = -b.DirX
		b.SpeedTimer -= b.SpeedTimer / 2
	} else {
		b.VZ = -b.VZ
		b.DirZ = -b.DirZ
	}
}
func warpBall(b *Ball, attribute int) {
	dx, dz := b.DirX, b.DirZ
	if dx == 0 && dz == 0 {
		dx, dz = eightWay(b.VX, b.VZ)
	}
	b.Z = -math.Copysign(11.2, b.Z)
	drift := b.VZ
	b.VX = dx * 8 * velocityUnit
	b.VZ = dz * 8 * velocityUnit
	if dz == 0 {
		b.VZ = drift
	}
	b.DirX, b.DirZ = dx, dz
	setBallSpeed(b, attribute)
}

// throwing_action_fn 0x10882: only a zero component can receive release input.
func steerRelease(b *Ball, input Input) {
	sign := func(v float64) float64 {
		if v == 0 {
			return 0
		}
		return math.Copysign(1, v)
	}
	if b.VZ == 0 {
		b.VZ = sign(input.Z) * 4 * velocityUnit
	} else if b.VX == 0 {
		b.VX = sign(input.X) * 4 * velocityUnit
	}
}

var highFlight = []int{1, 1, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 5, 5, 5, 4, 4, 4, 4, 4, 3, 3, 3, 3, 2, 2, 2, 1, 1, 0}

// Amiga anim_ball_launch at 0x6d8a.
var launchFlight = []int{0, 7, 7, 8, 8, 9, 9, 10, 10, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 10, 10, 9, 9, 8, 8, 7, 7, 0}
var lowFlight = []int{1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0}

func startFlight(b *Ball, high bool) {
	b.FlightKind = 1
	if high {
		b.FlightKind = 2
	}
	b.FlightIndex = -1
	b.FlightStage = 1
	b.FlightFraction = 0
	b.H = .75
	b.VH = 0
}
func flightStep(b *Ball, dt float64) bool {
	if b.FlightKind == 0 {
		return false
	}
	frames := lowFlight
	if b.FlightKind == 3 {
		frames = launchFlight
	}
	if b.FlightKind == 2 {
		frames = highFlight
	}
	b.FlightFraction += dt * 25
	for b.FlightFraction >= 1-1e-9 {
		b.FlightFraction = math.Max(0, b.FlightFraction-1)
		b.FlightIndex = min(b.FlightIndex+1, len(frames)-1)
		b.FlightStage = frames[b.FlightIndex]
	}
	heightStage := b.FlightStage
	if b.FlightKind == 3 && heightStage >= 7 {
		heightStage -= 6
	}
	b.H = .25 + float64(heightStage)*.5
	b.VH = 0
	return true
}
